homeassistant/components/rachio.py

```diff
@@ -0,0 +1,282 @@
+        # Save ID for deletion at shutdown
+        current_webhook_id = new_webhook[1][KEY_ID]
+
+        def _deinit_webhooks(event) -> None:
+            """Stop getting updates from the Rachio API."""
+            webhooks = self.rachio.notification.getDeviceWebhook(
+                self.controller_id)[1]
+            for webhook in webhooks:
+                if webhook[KEY_EXTERNAL_ID] == WEBHOOK_CONST_ID or\
+                        webhook[KEY_ID] == current_webhook_id:
+                    self.rachio.notification.deleteWebhook(webhook[KEY_ID])
+
+        self.hass.bus.listen(EVENT_HOMEASSISTANT_STOP, _deinit_webhooks)
+
+    def __str__(self) -> str:
+        """Display the controller as a string."""
+        return 'Rachio controller "{}"'.format(self.name)
+
+    @property
+    def controller_id(self) -> str:
+        """Return the Rachio API controller ID."""
+        return self._id
+
+    @property
+    def name(self) -> str:
+        """Return the user-defined name of the controller."""
+        return self._name
+
+    @property
+    def current_schedule(self) -> str:
+        """Return the schedule that the device is running right now."""
+        return self.rachio.device.getCurrentSchedule(self.controller_id)[1]
+
+    @property
+    def init_data(self) -> dict:
+        """Return the information used to set up the controller."""
+        return self._init_data
+
+    def list_zones(self, include_disabled=False) -> list:
+        """Return a list of the zone dicts connected to the device."""
+        # All zones
+        if include_disabled:
+            return self._zones
+
+        # Only enabled zones
+        return [z for z in self._zones if z[KEY_ENABLED]]
+
+    def get_zone(self, zone_id) -> dict or None:
+        """Return the zone with the given ID."""
+        for zone in self.list_zones(include_disabled=True):
+            if zone[KEY_ID] == zone_id:
+                return zone
+
+        return None
+
+    def stop_watering(self) -> None:
+        """Stop watering all zones connected to this controller."""
+        self.rachio.device.stopWater(self.controller_id)
+        _LOGGER.info("Stopped watering of all zones on %s", str(self))
+
+
+class RachioWebhookView(HomeAssistantView):
+    """Provide a page for the server to call."""
+
+    EVENT_TYPE_MAPPINGS = {
+        TYPE_CONTROLLER_STATUS: EVENT_RACHIO_CONTROLLER_UPDATE,
+        TYPE_SCHEDULE_STATUS: EVENT_RACHIO_SCHEDULE_UPDATE,
+        TYPE_ZONE_STATUS: EVENT_RACHIO_ZONE_UPDATE
+    }
+
+    requires_auth = True
+    url = WEBHOOK_PATH
+    name = url[1:].replace('/', ':')
+
+    # pylint: disable=no-self-use
+    @asyncio.coroutine
+    async def post(self, request) -> web.Response:
+        """Handle webhook calls from the server."""
+        data = await request.json()
+        update_type = data[KEY_TYPE]
+
+        if update_type in self.EVENT_TYPE_MAPPINGS:
+            request.app['hass'].bus.async_fire(
```

Comment:
Use our dispatch helper instead of the event bus to communicate between component and platforms.